File: java/ReturnKthToLast.java
// Implement an algorithm to find the kth to last element of a singly linked list


class ReturnKthToLast {
    public static void main(String args[]) {
        MyLinkedList<Integer> list = new MyLinkedList<Integer>(1);
        list.add(2);
        list.add(3);
        list.add(4);
        list.add(5);
        list.add(6);
        list.add(7);
        System.out.println(returnKthToLast(list, 3).data);
        System.out.println(returnKthToLastWithoutCount(list, 3).data);
    }

    // time complexity: O(n), space complexity: O(1)
    private static MyNode returnKthToLast(MyLinkedList list, Integer k) {
        int nodeIndex = list.length - k - 1;
        int currentIndex = 0;
        MyNode currentNode = list.head;
        while (currentNode.next != null && currentIndex != nodeIndex) {
            currentNode = currentNode.next;
            currentIndex += 1;
        }
        return currentNode;
    }

    // time complexity: O(n), space complexity: O(1)
    private static MyNode returnKthToLastWithoutCount(MyLinkedList list, Integer k) {
        MyNode currentNode = list.head;
        MyNode runner = list.head;
        for (int i = 0; i < k; i++) {
            runner = runner.next;
        }
        while (runner.next != null) {
            currentNode = currentNode.next;
            runner = runner.next;
        }
        return currentNode;
    }
}

File: python/return_kth_to_last.py
from linked_list import LinkedList

# Implement an algorithm to find the kth to last element of a singly linked list

# time complexity: O(n), space complexity: O(1)
def return_kth_to_last(list, k):
    node_index = list.count - k - 1
    current_node = list.head
    current_index = 0
    while current_node.next and current_index != node_index:
        current_node = current_node.next
        current_index += 1
    return current_node


# time complexity: O(n), space complexity: O(1)
def return_kth_to_last_without_count(list, k):
    current_node = list.head
    runner = list.head
    for i in range(k):
        runner = runner.next
    while runner.next != None:
        current_node = current_node.next
        runner = runner.next
    return current_node



list = LinkedList(1)
list.add(2)
list.add(3)
list.add(4)
list.add(5)
list.add(6)
list.add(7)

print(return_kth_to_last(list, 3).data)
print(return_kth_to_last_without_count(list, 3).data)
